File: backend/app/api/routes/media_imports.py
from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Request, Response, status

from app.api.admission import RateLimitAdmission
from app.api.auth_dependencies import get_current_user
from app.api.dependencies import (
    IdempotencyKey,
    MediaImportUseCases,
    get_media_import_use_cases,
    get_runtime_settings,
)
from app.api.errors import import_application_error
from app.api.schemas.media_imports import (
    CompleteMediaImportRequest,
    MediaImportRequest,
    MediaImportResponse,
    MediaUploadSessionResponse,
)
from app.api.upload_signing import use_local_browser_upload_endpoint
from app.application.auth import CurrentUser
from app.application.imports import CompletedUploadPart, ImportApplicationError
from app.domain.imports import ContentKind, ImportSourceFormat

router = APIRouter(prefix="/media-imports", tags=["media-imports"])
User = Annotated[CurrentUser, Depends(get_current_user)]
UseCases = Annotated[MediaImportUseCases, Depends(get_media_import_use_cases)]


@router.post(
    "",
    operation_id="createMediaImport",
    dependencies=[Depends(RateLimitAdmission("media_import"))],
    response_model=MediaImportResponse,
    status_code=status.HTTP_201_CREATED,
    summary="创建本地视频导入",
)
async def create_media_import(
    body: MediaImportRequest,
    idempotency_key: IdempotencyKey,
    response: Response,
    user: User,
    use_cases: UseCases,
) -> MediaImportResponse:
    """创建只接受 MP4 的浏览器上传资源，不接收任意存储参数。"""
    try:
        view = await use_cases.create_resource(
            owner_hash=user.owner_hash,
            idempotency_key=idempotency_key,
            content_kind=ContentKind.VIDEO,
            source_format=ImportSourceFormat.MP4,
            file_name=body.file_name,
            declared_size_bytes=body.declared_size_bytes,
            declared_sha256=body.declared_sha256,
            rights_accepted=body.rights_accepted,
            declared_origin=body.declared_origin,
        )
    except ImportApplicationError as error:
        raise import_application_error(error) from error
    response.headers["Location"] = f"/api/media-imports/{view.id}"
    response.headers["Cache-Control"] = "no-store"
    return MediaImportResponse.from_view(view)


@router.get(
    "/{resource_id}",
    operation_id="getMediaImport",
    response_model=MediaImportResponse,
    summary="查询本地视频导入",
)
async def get_media_import(
    resource_id: UUID,
    response: Response,
    user: User,
    use_cases: UseCases,
) -> MediaImportResponse:
    try:
        view = await use_cases.get_import(
            resource_id, user.owner_hash, ContentKind.VIDEO
        )
    except ImportApplicationError as error:
        raise import_application_error(error) from error
    response.headers["Cache-Control"] = "no-store"
    return MediaImportResponse.from_view(view)


@router.post(
    "/{resource_id}/upload-sessions",
    operation_id="createMediaUploadSession",
    dependencies=[Depends(RateLimitAdmission("media_import_upload"))],
    response_model=MediaUploadSessionResponse,
    status_code=status.HTTP_201_CREATED,
    summary="创建或刷新视频上传会话",
)
async def create_media_upload_session(
    resource_id: UUID,
    request: Request,
    response: Response,
    user: User,
    use_cases: UseCases,
) -> MediaUploadSessionResponse:
    try:
        view = await use_cases.create_upload_session(
            resource_id,
            user.owner_hash,
            ContentKind.VIDEO,
            use_local_browser_endpoint=use_local_browser_upload_endpoint(
                request, get_runtime_settings(request)
            ),
        )
    except ImportApplicationError as error:
        raise import_application_error(error) from error
    response.headers["Cache-Control"] = "no-store"
    return MediaUploadSessionResponse.from_view(view)


@router.post(
    "/{resource_id}/complete",
    operation_id="completeMediaImport",
    dependencies=[Depends(RateLimitAdmission("media_import_upload"))],
    response_model=MediaImportResponse,
    summary="完成视频上传并触发验证",
)
async def complete_media_import(
    resource_id: UUID,
    body: CompleteMediaImportRequest,
    response: Response,
    user: User,
    use_cases: UseCases,
) -> MediaImportResponse:
    try:
        view = await use_cases.complete_upload(
            resource_id,
            user.owner_hash,
            ContentKind.VIDEO,
            tuple(
                CompletedUploadPart(part.part_number, part.etag) for part in body.parts
            ),
        )
    except ImportApplicationError as error:
        raise import_application_error(error) from error
    response.headers["Cache-Control"] = "no-store"
    return MediaImportResponse.from_view(view)

File: frontend/src/api/mediaImports.ts
// @ts-ignore
/* eslint-disable */
import { request, type RequestOptions } from "@/lib/request";

/** 创建本地视频导入 创建只接受 MP4 的浏览器上传资源，不接收任意存储参数。 POST /api/media-imports */
export async function createMediaImport(
  body: API.MediaImportRequest,
  options?: RequestOptions
) {
  return request<API.MediaImportResponse>("/api/media-imports", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** 查询本地视频导入 GET /api/media-imports/${param0} */
export async function getMediaImport(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getMediaImportParams,
  options?: RequestOptions
) {
  const { resource_id: param0, ...queryParams } = params;
  return request<API.MediaImportResponse>(`/api/media-imports/${param0}`, {
    method: "GET",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** 完成视频上传并触发验证 POST /api/media-imports/${param0}/complete */
export async function completeMediaImport(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.completeMediaImportParams,
  body: API.CompleteMediaImportRequest,
  options?: RequestOptions
) {
  const { resource_id: param0, ...queryParams } = params;
  return request<API.MediaImportResponse>(
    `/api/media-imports/${param0}/complete`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      params: { ...queryParams },
      data: body,
      ...(options || {}),
    }
  );
}

/** 创建或刷新视频上传会话 POST /api/media-imports/${param0}/upload-sessions */
export async function createMediaUploadSession(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.createMediaUploadSessionParams,
  options?: RequestOptions
) {
  const { resource_id: param0, ...queryParams } = params;
  return request<API.MediaUploadSessionResponse>(
    `/api/media-imports/${param0}/upload-sessions`,
    {
      method: "POST",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}
